### PR TITLE
[MRG] ENH: Function to get spike rates

### LINKED ISSUE #154: calculate and plot average spike rate of each cell types
the average spike rate can be used as a criteria for rejecting a parameter set in the spiking is out of physiological ranges.

### PR DESCRIPTION
**Pull request by jasmainak**
closes #154 

not sure if this is the right way to do it, just opening a "work in progress thing" and we can discuss the details. How is the spike rate typically calculated?

TODO

- [x] tests
- [x] example
- [x] what's new page

### CODE REVIEW

File: hnn_core/network.py [hunk 1]
```diff
@@ -466,6 +466,27 @@ def update_types(self, gid_dict):
...
+
+        Returns
+        -------
+        spike_rate : dict
+            Dictionary with keys 'L5_pyramidal', 'L5_basket', etc.
+        """
+        cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+        spike_rates = dict()
+        all_spike_times = np.array(sum(self._times, []))
+        tstart, tstop = min(all_spike_times), max(all_spike_times)
```
Comment (jasmainak) [inline hnn_core/network.py:480]: one complication that creeps up here is that we don't have access to the start and end time of the trials in the spike object. Perhaps that needs to be an attribute of the spikes object. Or perhaps we need another "times" array which stores the time instances of simulation and a different "spike_times" arrange which tells us when in the "times" array we get a spike.

**Comment by jasmainak** [comment]
@ntolley I think you should take over this PR after the other one is merged. You've been working the spikes object recently, so this is right up your alley. Just add my fork as a remote on git and start pushing to the branch. Let me know if you need help.

**Comment by ntolley** [comment]
That sounds great to me! Perhaps not for this initial PR, but something to consider in the future is adding a convolution kernel to calculate spike rates. [Here](https://elephant.readthedocs.io/en/latest/reference/toctree/statistics/elephant.statistics.instantaneous_rate.html#elephant.statistics.instantaneous_rate) is a great example of how that sort function would work.

**Comment by jasmainak** [comment]
As long as you don't create a dependency on Neo, feel free to do it in this PR :-) Just copy the relevant code (they are BSD-3) and keep it simple. You can have an argument called `'method'`

**Comment by ntolley** [comment]
Reading the issue more closely it  actually seems like convolved spike rates gets at a different purpose. 

Estimation of instantaneous rates is usually done at the single unit level, whereas @stephanie-r-jones  proposes aggregating across cell types to describe the whole trial. 

In my mind these are rather distinct and may be better split into two functions. `mean_cell_rate()` could return mean firing rate (of each trial separately?) across cells. Then `spike_rate()` could implement instantaneous rate estimation for each cell separately, returning an array of convolved rates indexed by gid. Still not obvious to me if data from each trial should be kept separate.

**Comment by jasmainak** [comment]
humm ... do you think the mean spiking rates (across trials and cells) could be added to this plot: https://jonescompneurolab.github.io/hnn-core/generated/hnn_core.viz.plot_spikes_raster.html#hnn_core.viz.plot_spikes_raster instead of a function since it's just 4 scalar values?

and we can also have another function for instantaneous rate, but I'm wondering if that should be reserved for another PR? What's not totally obvious to me in that case is:
a) what's the use case for this as opposed to the just mean spiking rates
b) what would the function signature look like and what would be the data structure of what the function returns? I'm assuming you'd need a parameter to control the width of a convolution kernel? And the returned data would be an array of shape `n_cells x n_trials` ?

**Comment by ntolley** [comment]
That could possibly work, but it doesn't lend itself easily to accessing the data. This may be a good opportunity to create an `analysis.py` module to give us more flexibility with feature requests in the future. If it seems important enough to create a member function on the `Spikes` class, we can just call `analysis.py` in the same way `viz.py` is implemented.

wrt plotting I think a box-plot makes the most sense for visualizing the rates across cells so perhaps we can append a subplot to the raster.

For the use case I personally intend to analyze single unit activity in my own projects. Generally it will be extremely useful to have available for highly mechanistic hypotheses.  And yes the function would generally look like `spike_rates(spikes, kernel='exponential', width=0.1, gids='all')`. A matrix return could work, it could also return a dictionary indexed by gid.

**Comment by jasmainak** [comment]
Let's worry about `analysis` module later :) I tend to move things into a module once there are enough features to fit in there. If you plan for "future", it will lead to an overengineered solution that is hard to maintain.

So, if I understand correctly are you leaning towards:

1) function for computing instantaneous spike rate?
2) a box plot to go with it

If the user wanted average spike rates across cell types, how would they get it? would they average the return value from that function? or do you want another function for that? or should they just look at the plot which will have the values?

**Comment by ntolley** [comment]
Honestly I think keeping these separate is the best path. For this PR I'm leaning towards a function to return mean spike rate grouped by cell. Example return would be `{'L5_basket: [trial1_mean, trial2_mean], ...}`.  Since the issue raised specifically refers to using this as an optimization heuristic it should be easy to access the data. 

To be clear, for this PR there would be 2 additions to the `Spikes` class: `mean_rate()` and `plot_mean_rate()`. Perhaps eliminate "spike_rate" naming convention since it lives in the Spikes class?

After that I can open a separate PR to introduce `instantaneous_rate()`.

**Comment by jasmainak** [comment]
okay, let's start coding :) We can iterate the API once we have something down. Can you push to this branch?

**Comment by ntolley** [comment]
Sorry it looks like this branch was behind your master so the rebase included the latest PR's merged. I'll work on cleaning it up tomorrow morning. 

In either case all my work is on the last commit. It now computes the mean spike rates for every individual gid, you can then specify what type of average you want returned by setting `mean_type='all'`, `'trial'`, or `'cell'`.

File: hnn_core/network.py [hunk 2]
```diff
@@ -459,14 +469,34 @@ def get_spike_rates(self):
...
+            gid_spike_rate = np.zeros((len(self._times), len(cell_type_gids)))
+
+            trial_data = zip(self._types, self._gids)
+            for trial, (spike_types, spike_gids) in enumerate(trial_data):
+                trial_type_mask = np.in1d(spike_types, cell_type)
+                gid, gid_counts = np.unique(np.array(
+                    spike_gids)[trial_type_mask], return_counts=True)
+
+                gid_spike_rate[trial, cell_type_gids == gid] = (gid_counts / (
+                    tstop - tstart)) * 1000
```
Comment (jasmainak) [inline hnn_core/network.py:488]: we still need to figure out where to get the `tstart` and `tstop` from. Should it be part of Spikes object or an argument to this method?

File: hnn_core/network.py [hunk 3]
```diff
@@ -459,14 +469,34 @@ def get_spike_rates(self):
...
-                spike_times_cells = spike_times[spike_types == cell_type]
-                trial_spike_rate.append(
-                    len(spike_times_cells) / (tstop - tstart))
-            spike_rates[cell_type] = np.mean(trial_spike_rate)
+            type_mask = np.in1d(all_spike_types, cell_type)
+            cell_type_gids = np.unique(all_spike_gids[type_mask])
+            gid_spike_rate = np.zeros((len(self._times), len(cell_type_gids)))
+
+            trial_data = zip(self._types, self._gids)
+            for trial, (spike_types, spike_gids) in enumerate(trial_data):
```
Comment (jasmainak) [inline hnn_core/network.py:482]:
```suggestion
            for trial_idx, (spike_types, spike_gids) in enumerate(trial_data):
```

I always use `_idx` whenever it is an index.

File: hnn_core/network.py [hunk 4]
```diff
@@ -459,14 +469,34 @@ def get_spike_rates(self):
         cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
         spike_rates = dict()
         all_spike_times = np.array(sum(self._times, []))
+        all_spike_types = np.array(sum(self._types, []))
+        all_spike_gids = np.array(sum(self._gids, []))
```
Comment (jasmainak) [inline hnn_core/network.py:473]: something seems complicated with the logic. I can't quite pinpoint it without touching the code but it seems to me that you first flatten across trials and then do a for loop to access trial-specific info which makes it complicated.

**Comment by jasmainak** [comment]
Good luck with the rebase. Maybe the force be with you.

You can go ahead and add the tests as well.

**Comment by ntolley** [comment]
I've taken a chainsaw to the commit tree and it's looking beautiful again.

Regarding the tests, obvious ones are checking for correct `mean_type` calls. Since this relies on data members of the `Spikes` class I believe the existing tests will catch any problems that could arise in the rate calculations. Can you think of anything that should be caught?

File: hnn_core/network.py (hunk 4 above)
Comment (ntolley) [reply hnn_core/network.py:473]: Since I reworked the rate calculation to start with individual gids, the flattening allows you to pre-allocate space for every gid. The rates are stored in the array `gid_spike_rate` with dimensions (trial x gid). This deals with the situation where specific gids don't spike on every trial; their inactivity should be included in the average across cells tyoes.

I'm sure there's a cleaner way to achieve this but that's the rationale.

File: hnn_core/network.py (hunk 2 above)
Comment (ntolley) [reply hnn_core/network.py:488]: As I think about it more it definitely makes sense for the `Spikes` to store this information. The interpretation of a spike train is very different if you can't differentiate inactivity from the recording being cut short.

Perhaps it can be an added to the instantiation as `self._bounds = (t_start, t_stop)`?

**Comment by jasmainak** [comment]
For tests, you should create a `Spikes` object where you know what the spike rate is and then check that the function returns the same spike rate.

**Comment by jasmainak** [comment]
> I've taken a chainsaw to the commit tree and it's looking beautiful again.

amazing!! :)

File: hnn_core/network.py (hunk 2 above)
Comment (jasmainak) [reply hnn_core/network.py:488]: I would just do `self.tstart` and `self.tstop` and then use that in the spike rate calculation. You'll also need to add tests to check they are set correctly when `Spikes` is an attribute of `Network`.

File: hnn_core/network.py (hunk 4 above)
Comment (jasmainak) [reply hnn_core/network.py:473]: I'll take a look at this later. My brain doesn't want to think on a Saturday evening :) 

Do remember to update an example and also to update the "Attributes" section of the `Spikes` class once you add `tstart` and `tstop`

**Comment by ntolley** [comment]
Surprisingly beefy commit. I've implemented the tstart and tstop attributes. Let me know if you disagree, but the it seemed like the most logical place to instantiate the attributes was `_gather_trial_data()` so all the data is stored from the same function.

Also there is now an issue with legacy `spk_*.txt` files and storing the new attributes; the write methods now include tstart and tstop. I've added additional logic to check for these additional columns.

TODO:
- [ ]  Update examples
- [ ] Add test for precomputed rates

File: hnn_core/network.py [hunk 5]
```diff
@@ -345,6 +376,8 @@ class Spikes(object):
         The inner list contains the type of spike (e.g., evprox1
         or L2_pyramidal) that occured at the corresonding time stamp.
         Each gid corresponds to a type via Network::gid_dict.
+    tstart, tstop : int | float | None
+        Values defining the start and stop times of all trials.
```
Comment (jasmainak) [inline hnn_core/network.py:380]: document tstart and tstop in 2 separate lines just to conform to the conventions

File: hnn_core/network.py [hunk 6]
```diff
@@ -448,6 +497,78 @@ def update_types(self, gid_dict):
             spike_types += [list(spike_types_trial)]
         self._types = spike_types
 
+    def update_trial_bounds(self, tstart, tstop):
```
Comment (jasmainak) [inline hnn_core/network.py:500]: humm ... I don't quite see the value of this method. Do we see any usecase for the user to modify the `tstart` and `tstop`?

File: hnn_core/tests/test_network.py [hunk 7]
```diff
@@ -96,16 +110,32 @@ def test_spikes():
     with pytest.raises(ValueError, match="No input types found for ABC"):
         spikes.plot_hist(spike_types='ABC', show=False)
 
+    with pytest.raises(ValueError, match="Invalid mean_type. Valid "
+                       "arguments include 'all', 'trial', or 'cell'."):
+        spikes.mean_rates(mean_type='ABC')
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:115]: can we have more thorough tests for `mean_rates`? Like have some simulated spikes and check that the spike rate comes out to be what you expected. Try different corner cases, e.g., when one trial has no spikes -- does the mean come out correctly etc.

File: hnn_core/network.py [hunk 8]
```diff
@@ -35,27 +37,54 @@ def read_spikes(fname, gid_dict=None):
...
         spike_trial = np.loadtxt(file, dtype=str)
         spike_times += [list(spike_trial[:, 0].astype(float))]
         spike_gids += [list(spike_trial[:, 1].astype(int))]
 
         # Note that legacy HNN 'spk.txt' files don't contain a 3rd column for
-        # spike type. If reading a legacy version, validate that a gid_dict is
-        # provided.
-        if spike_trial.shape[1] == 3:
+        # spike type, or a 4th and 5th column for tstart and tstop. If reading
+        # a legacy version, validate that gid_dict, tstart, and tstop provided.
```
Comment (jasmainak) [inline hnn_core/network.py:49]: why another column? I would have added it in the beginning before starting the rows since it's going to be the same for all spikes. See also related discussion [here](https://github.com/jonescompneurolab/hnn-core/pull/83#discussion_r478705027) and in #159

File: hnn_core/network.py [hunk 9]
```diff
@@ -35,27 +37,54 @@ def read_spikes(fname, gid_dict=None):
...
+    if len(np.unique(spike_tstart)) > 1 or len(np.unique(spike_tstop)) > 1:
+        raise ValueError("tstart and tstop must match across files.")
+
+    elif len(np.unique(spike_tstart)) == 1 and len(
+            np.unique(spike_tstop)) == 1:
+        spike_tstart = spike_tstart[0]
+        spike_tstop = spike_tstop[0]
+    else:
+        spike_tstart = np.min(sum(spike_times, []))
+        spike_tstop = np.max(sum(spike_times, []))
```
Comment (jasmainak) [inline hnn_core/network.py:75]: I would say we shouldn't guess if the values are not given. It would give incorrect results, no?

Also, given [L62-64](https://github.com/jonescompneurolab/hnn-core/pull/155/files#diff-3e7f4ea1f282c3abd4802eef071e9a2aR62-R64), is there any chance of hitting this `else` condition?

**Review by jasmainak — COMMENTED** [review]
It's converging @ntolley slowly but surely :)

File: hnn_core/network.py (hunk 9 above)
Comment (ntolley) [reply hnn_core/network.py:75]: Definitely a brain fart there, I agree this won't be reached. I'll convert the preceding `elif` clause to `else`.

File: hnn_core/network.py (hunk 6 above)
Comment (ntolley) [reply hnn_core/network.py:501]: Perhaps this is an excessive solution, but it's added to work with the `read_spikes()` method [here](https://github.com/jasmainak/hnn-core/blob/c770df17400e56ab4ea7fd676adca0d8aed9da43/hnn_core/network.py#L81). This allows for the tstart/tstop tests to live in the same place; similar to `gid_dict` there are different ways it gets instantiated to allow for older file versions. 

To clarify this copies the way `update_types()` is used [here](https://github.com/jasmainak/hnn-core/blob/c770df17400e56ab4ea7fd676adca0d8aed9da43/hnn_core/network.py#L78). I am assuming method is really just used for error raising since the return on `read_spikes()` is a separate object? Perhaps that is a mistake that wasn't caught earlier...

File: hnn_core/network.py (hunk 8 above)
Comment (ntolley) [reply hnn_core/network.py:49]: The main motivation was to make it easy to check the version of the `spk.txt` file (since there are already a variable number of columns between the versions). 

What do you think of adding a string to the top left entry that will indicate fixed parameters. Something like `f.write('{}\t{:.3f}\t{:.3f}\n'.format('Fixed Values', float(self._tstart), 'float(self._tstop)')`?

That way the logic just involved checking the string and skipping rows if it is present.

File: hnn_core/network.py [hunk 10]
```diff
@@ -25,6 +25,8 @@ def read_spikes(fname, gid_dict=None):
         containing the range of Cell or input IDs of different
         cell or input types. If None, each spike file must contain
         a 3rd column for spike type.
+    tstart, tstop : float | None
```
Comment (jasmainak) [inline hnn_core/network.py:28]: still need to document in two different lines

File: hnn_core/network.py (hunk 8 above)
Comment (jasmainak) [reply hnn_core/network.py:49]: I was thinking of using [`np.loadtxt`](https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html) with comments="#". Then you can have two lines in the beginning which specify tstart and tstop. But how will this work for multiple trials? Are we going to check that tstart and tstop is same across trials? Will each trial be stored in different files or the same?

I feel we shouldn't develop a new format ad hoc and spend some time to think what should be the ideal format and what are the pros and cons of different formats. Probably a good discussion topic for an HNN meeting or even during the hackathon.

Comment (jasmainak) [reply hnn_core/network.py:49]: Tagging @blakecaldwell and @rythorpe for suggestions

File: hnn_core/network.py [hunk 11]
```diff
@@ -35,27 +37,51 @@ def read_spikes(fname, gid_dict=None):
...
                 raise ValueError("gid_dict must be provided if spike types "
                                  "are unspecified in the file %s" % (file,))
             spike_types += [[]]
 
-    spikes = Spikes(times=spike_times, gids=spike_gids, types=spike_types)
+        if spike_trial.shape[1] == 5:
+            spike_tstart += np.unique(spike_trial[:, 3].astype(float)).tolist()
+            spike_tstop += np.unique(spike_trial[:, 4].astype(float)).tolist()
+        else:
+            if tstart is None or tstop is None:
```
Comment (jasmainak) [inline hnn_core/network.py:62]: this should be an elif, no? I don't see why the if should be nested

File: hnn_core/network.py [hunk 12]
```diff
@@ -35,27 +37,51 @@ def read_spikes(fname, gid_dict=None):
...
-        else:
+        elif spike_trial.shape[1] == 2:
             if gid_dict is None:
                 raise ValueError("gid_dict must be provided if spike types "
                                  "are unspecified in the file %s" % (file,))
             spike_types += [[]]
 
-    spikes = Spikes(times=spike_times, gids=spike_gids, types=spike_types)
+        if spike_trial.shape[1] == 5:
+            spike_tstart += np.unique(spike_trial[:, 3].astype(float)).tolist()
```
Comment (jasmainak) [inline hnn_core/network.py:59]: I know it's not your code but can you see if the `astype(float)` can be avoided by using the `dtype` argument in `np.loadtxt`?

File: hnn_core/network.py [hunk 13]
```diff
@@ -35,27 +37,51 @@ def read_spikes(fname, gid_dict=None):
...
+            spike_tstop += np.unique(spike_trial[:, 4].astype(float)).tolist()
+        else:
+            if tstart is None or tstop is None:
+                raise ValueError("tstart and tstop must be provided if values "
+                                 "are unspecified in the file %s" % (file,))
+
+    if len(np.unique(spike_tstart)) > 1 or len(np.unique(spike_tstop)) > 1:
+        raise ValueError("tstart and tstop must match across files.")
```
Comment (jasmainak) [inline hnn_core/network.py:67]: but do you check somewhere that it matches within a file?

File: hnn_core/network.py [hunk 14]
```diff
@@ -25,6 +25,8 @@ def read_spikes(fname, gid_dict=None):
         containing the range of Cell or input IDs of different
         cell or input types. If None, each spike file must contain
         a 3rd column for spike type.
+    tstart, tstop : float | None
+        Values defining the start and stop times of all trials.
```
Comment (jasmainak) [inline hnn_core/network.py:29]: can you also say that this is only relevant for legacy files? You can also add a section below "Returns", call it "Notes" and describe the file format as in here: https://github.com/mne-tools/mne-python/blob/master/mne/channels/montage.py#L480-L521

File: hnn_core/network.py [hunk 15]
```diff
@@ -35,27 +37,51 @@ def read_spikes(fname, gid_dict=None):
...
+        elif spike_trial.shape[1] == 2:
             if gid_dict is None:
                 raise ValueError("gid_dict must be provided if spike types "
                                  "are unspecified in the file %s" % (file,))
             spike_types += [[]]
 
-    spikes = Spikes(times=spike_times, gids=spike_gids, types=spike_types)
+        if spike_trial.shape[1] == 5:
+            spike_tstart += np.unique(spike_trial[:, 3].astype(float)).tolist()
+            spike_tstop += np.unique(spike_trial[:, 4].astype(float)).tolist()
```
Comment (jasmainak) [inline hnn_core/network.py:60]: if the user provides a `tstart` / `tstop` here it will be silently ignored which could be surprising. What's the desired behavior? Should you raise an error? Or always treat the user provided one with higher priority?

**Comment by jasmainak** [comment]
Sorry @ntolley this PR turned out to be more complex than I anticipated. Don't lose hope!! we'll get it merged soon :)

File: hnn_core/network.py (hunk 6 above)
Comment (jasmainak) [reply hnn_core/network.py:501]: I see, that makes sense, but I think we should try to [keep it simple stupid](https://en.wikipedia.org/wiki/KISS_principle). A good rule of thumb is that you need a function only if something is used more than once. It seems [update_types](https://github.com/jonescompneurolab/hnn-core/search?q=update_types&unscoped_q=update_types) is, so it looks fair. I would suggest moving the contents of this method directly into `read_spikes`

File: hnn_core/network.py (hunk 15 above)
Comment (ntolley) [reply hnn_core/network.py:67]: I agree, it's not clear to me if overriding information from the file is a realistic use case. If the user really cares they can edit the class post-hoc. I've added a test to raise an error for now.

File: hnn_core/network.py (hunk 14 above)
Comment (ntolley) [reply hnn_core/network.py]: Thanks for linking this, I'll hold off on describing the file format until we've settled on a consistent way to store fixed variables. I do like the way the comment lines are implemented here though, especially since it allows us to add more information in future PR's without too many nightmares with legacy format.

File: hnn_core/network.py (hunk 12 above)
Comment (ntolley) [reply hnn_core/network.py:66]: It looks like `np.loadtxt()` normally accepts a single input for `dtype`, in this case it's set to `str`. There is the option for more complex arguments but you need to know the file structure beforehand. We could potentially have two calls loading specific columns but I don't necessarily see the benefit. 

Do you mind explaining the issue with recasting to `float`'s?

File: hnn_core/tests/test_network.py (hunk 7 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py:133]: Looks like the manually written `spikes` object in `test_network.py` has this exact corner case (2 trials, each cell only spikes once per trial). I've added tests asserting the correct calculation for all three `mean_types`. I'll try and think of trickier edge cases before merging but this seems sufficient for now.

**Comment by jasmainak** [comment]
@ntolley I will get to this soon. Feel free to ping me if I don't in a few days :)

File: hnn_core/network.py (hunk 8 above)
Comment (rythorpe) [reply hnn_core/network.py:53]: We have needed to decide on a way to provide annotations in hnn-core data files for awhile, so I agree with @jasmainak that this might be a good time to start thinking about how to do this systematically. I'm unfamiliar with best-practices in this area, but I like the idea of adding annotations as lines to the top of each file. We could even supply a `legacy` boolean argument to the `read_spikes()` function in order to simplify the logic that handles identification of current vs. legacy versions.

File: hnn_core/network.py (hunk 12 above)
Comment (rythorpe) [reply hnn_core/network.py:66]: I had a similar issue when creating the write functions since there is a mixture of strings and numbers. I agree that specifying the `dtype` argument should be used from within `np.loadtxt()` whenever possible, but I'm not sure of an efficient way to use it here.

File: hnn_core/network.py (hunk 15 above)
Comment (rythorpe) [reply hnn_core/network.py:67]: Maybe the best way to handle this is to only raise an error if 1) tstart and tstop are provided in both the file and function call and 2) they are inconsistent.

File: hnn_core/network.py [hunk 16]
```diff
@@ -448,6 +508,82 @@ def update_types(self, gid_dict):
...
+        all_spike_types = np.array(sum(self._types, []))
+        all_spike_gids = np.array(sum(self._gids, []))
+
+        if mean_type not in ['all', 'trial', 'cell']:
+            raise ValueError("Invalid mean_type. Valid arguments include "
+                             "'all', 'trial', or 'cell'.")
+
+        for cell_type in cell_types:
+            type_mask = np.in1d(all_spike_types, cell_type)
+            cell_type_gids = np.unique(all_spike_gids[type_mask])
```
Comment (rythorpe) [inline hnn_core/network.py:563]: It looks like you only consider only the gids that spike somewhere in at least one trial rather than *all* gids of a given type. For example, one might find that only 2 L5 pyramidal cells fire out of the entire network of L5 pyramidal cells, across trials. Thus, I think your code here will report an average spike rate across cells as the average of the 2 cells that fired rather than the 100 L5 pyramidal cells in the network that receive an input. Is this correct?

Comment (rythorpe) [reply hnn_core/network.py:563]: In order to rectify this, you'll probably need to pass in the `Network.gid_dict` attribute. I'm beginning to wonder if this function belongs in the hnn_core analysis module.

Comment (ntolley) [reply hnn_core/network.py:563]: Great catch Ryan! That is correct, the mean rates would be incorrectly calculated when averaging across trials.

Also I may have had the same thought at [one point](https://github.com/jonescompneurolab/hnn-core/pull/155#issuecomment-684958576) ;)

In either case I think using `Network.gid_dict` would greatly simplify the calculations. At the outset the function could make a matrix that contains rates for every possible gid and trial. Then averaging would just require filtering by the gid index.

Comment (jasmainak) [reply hnn_core/network.py:563]: another alternative is to have `gid_dict` as part of the `Spikes` class, although I'm not sure how you would save it to files then ...

Comment (ntolley) [reply hnn_core/network.py:563]: If we move forward with placing attributes as comments at the top, it wouldn't be too unreasonable to add `gid_dict` directly since it's just a list of cell types and ranges. 

The two major decisions at the moment are: 
- [ ] Decide on file format for saving comments at the top (applies to the other `write` functions as well)
- [ ] Keep this function in the `Spikes` class or move it to an analysis module

I'm going to hold off on pushing more commits until that's settled. Let me know if you think both answers are pretty clear or if they warrant more discussion.

Comment (jasmainak) [reply hnn_core/network.py:563]: thinking again, I think `gid_dict` should just be passed to the function ... it can be `None` by default but must be a `dict` if you want to average by cell type. Same with `tstart` and `tstop` which should be required arguments. So the user has to be a bit careful about keeping track of these. But that way we don't worry about saving these information from `Spikes` object and can hold off the file format discussion for later and can move on with this PR.

as for the function, keep it in here for now. But let's keep a running record of what needs to go into the `analysis` module. Maybe in #139 ? When we have enough functions, we can move stuff there.

File: hnn_core/tests/test_network.py [hunk 17]
```diff
@@ -62,23 +62,55 @@ def test_spikes():
...
     spiketypes = [['L2_pyramidal', 'L2_basket'], ['L5_pyramidal', 'L5_basket']]
-    spikes = Spikes(times=spiketimes, gids=spikegids, types=spiketypes)
+    spikes = Spikes(times=spiketimes, gids=spikegids, types=spiketypes,
+                    tstart=0.1, tstop=98.4)
     spikes.plot_hist(show=False)
     spikes.write('/tmp/spk_%d.txt')
     assert spikes == read_spikes('/tmp/spk_*.txt')
     assert ("Spikes | 2 simulation trials" in repr(spikes))
+    assert spikes.mean_rates() == {
+        'L5_pyramidal': 5.08646998982706,
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:72]: do you mind explaining how you got this number given the numbers above?

Comment (ntolley) [reply hnn_core/tests/test_network.py:72]: Perhaps a janky solution, it's manually computed with numpy. On the next commit I can just use `np.isclose()` if you agree.

File: hnn_core/network.py (hunk 16 above)
Comment (ntolley) [reply hnn_core/network.py:563]: That sounds good to me. I'll start working on this and remove the modifications to the `write` function.

Comment (ntolley) [reply hnn_core/network.py]: Given @rythorpe's comment, I think you need an account of every gid for all calculations. The only exception would be making the default behavior return mean rates for each individual gid with `mean_types='cell'`.

The current commit requires a gid_dict in all cases.

File: hnn_core/network.py (hunk 4 above)
Comment (ntolley) [reply hnn_core/network.py]: No longer applicable to current commit.

File: hnn_core/tests/test_network.py (hunk 17 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py]: I've replaced this with an explicit calculation of the rate in `test_network.py` [here](https://github.com/jasmainak/hnn-core/blob/d69566307e86aaa886d80fb62b327e62f9d6f98d/hnn_core/tests/test_network.py#L143).

Do you think this is a good solution? One ugly alternative is looping through each dictionary and asserting `np.allclose()` on each pass.

**Comment by ntolley** [comment]
Definitely procrastinated on rebasing for too long. In line with the previous discussion I removed all of the code related to storing `tstart` and `tstop` as attributes and writing to a file. Instead they are passed to `mean_rates()` as parameters along with `gid_dict`. This additionally fixes the case where gids that never spike across trials aren't included in the final calculation.

This heavily simplifies the code and cleans up the awkward logical checks.

File: doc/whats_new.rst [hunk 18]
```diff
@@ -29,6 +29,8 @@ Changelog
 
 - Update plot_hist_input() to plot_spikes_hist() which can plot histogram of spikes for any cell type, by `Nick Tolley`_ in `#157 <https://github.com/jonescompneurolab/hnn-core/pull/157>`_
 
+- Add function to compute mean spike rates with user specified calculation type, by `Nick Tolley` and `Mainak Jas`_ in `#155 <https://github.com/jonescompneurolab/hnn-core/pull/155>`_
```
Comment (jasmainak) [inline doc/whats_new.rst:32]: your name won't have the hyperlink. Do you know how to build the documentation locally? You need to run:

$ make no-html

from inside the doc/ folder

File: hnn_core/network.py [hunk 19]
```diff
@@ -12,7 +12,7 @@
 from .viz import plot_spikes_hist, plot_spikes_raster, plot_cells
 
 
-def read_spikes(fname, gid_dict=None):
+def read_spikes(fname, gid_dict=None, tstart=None, tstop=None):
```
Comment (jasmainak) [inline hnn_core/network.py:15]: cruft from before?

File: hnn_core/network.py [hunk 20]
```diff
@@ -431,6 +434,71 @@ def update_types(self, gid_dict):
...
+        -------
+        spike_rate : dict
+            Dictionary with keys 'L5_pyramidal', 'L5_basket', etc.
+        """
+        cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+        spike_rates = dict()
+
+        if mean_type not in ['all', 'trial', 'cell']:
+            raise ValueError("Invalid mean_type. Valid arguments include "
+                             "'all', 'trial', or 'cell'.")
```
Comment (jasmainak) [inline hnn_core/network.py:468]: It's usually also a good idea to say what you got
```suggestion
                             f"'all', 'trial', or 'cell'. Got {mean_type}")
```

File: hnn_core/network.py [hunk 21]
```diff
@@ -431,6 +434,71 @@ def update_types(self, gid_dict):
...
+        # Validate tstart, tstop
+        if not isinstance(tstart, (int, float)) or not isinstance(
+                tstop, (int, float)):
+            raise ValueError('tstart and tstop must be of type int or float')
+        elif tstop <= tstart:
+            raise ValueError('tstop must be greater than tstart')
+
+        for cell_type in cell_types:
+            cell_type_gids = np.array(gid_dict[cell_type])
+            gid_spike_rate = np.zeros((len(self._times), len(cell_type_gids)))
```
Comment (jasmainak) [inline hnn_core/network.py:479]: Reading the code it was not obvious to me that `len(self._times)` is actually number of trials. I would add one line above to make it obvious
```suggestion
            n_trials, n_cells = len(self._times), len(cell_type_gids)
            gid_spike_rate = np.zeros((n_trials, n_cells))
```

File: hnn_core/network.py [hunk 22]
```diff
@@ -431,6 +434,71 @@ def update_types(self, gid_dict):
...
+                trial_type_mask = np.in1d(spike_types, cell_type)
+                gids, gid_counts = np.unique(np.array(
+                    spike_gids)[trial_type_mask], return_counts=True)
+
+                gid_spike_rate[trial_idx, np.in1d(cell_type_gids, gids)] = (
+                    gid_counts / (tstop - tstart)) * 1000
+
+            if mean_type == 'all':
+                spike_rates[cell_type] = np.mean(
+                    np.mean(gid_spike_rate, axis=1))
```
Comment (jasmainak) [inline hnn_core/network.py:492]: you can just do `gid_spike_rate.mean()`. Simpler and also easier to read ...

File: hnn_core/network.py [hunk 23]
```diff
@@ -431,6 +434,71 @@ def update_types(self, gid_dict):
...
+            if mean_type == 'trial':
+                spike_rates[cell_type] = np.mean(
+                    gid_spike_rate, axis=1).tolist()
+            if mean_type == 'cell':
+                spike_rates[cell_type] = [gid_trial_rate.tolist()
+                                          for gid_trial_rate in gid_spike_rate]
```
Comment (jasmainak) [inline hnn_core/network.py:498]: I think one issue here is that the user will not know which cell the spike rate belongs to? does it matter? If it does, you might want to consider returning a dictionary with the `gid` as key instead of `cell_type`

**Comment by jasmainak** [comment]
@ntolley looks pretty clean to me and see you have added tests too! Fantastic.

Would you mind addressing the last few comments and then we can merge. Don't forget to change the title from WIP to MRG once you are ready

File: hnn_core/network.py [hunk 24]
```diff
@@ -500,7 +568,7 @@ def write(self, fname):
         for trial_idx in range(len(self._times)):
             with open(str(fname) % (trial_idx,), 'w') as f:
                 for spike_idx in range(len(self._times[trial_idx])):
-                    f.write('{:.3f}\t{}\t{}\n'.format(
+                    f.write('{:.3f}\t{}\t{}\t\n'.format(
```
Comment (jasmainak) [inline hnn_core/network.py:571]: was this intentional?

**Comment by jasmainak** [comment]
Oh, one last thing is that can you update an example to show the use of this new function?

File: hnn_core/network.py (hunk 23 above)
Comment (ntolley) [reply hnn_core/network.py:498]: Since the user is currently required to pass `gid_dict` the return type will have an identical structural here. The difference is the gid values will be replaced with rates. 

I think the more common use case will be analyzing the distribution of rates as grouped by cell type. In either situation it's a small matter of index filtering to convert between these option.

File: hnn_core/network.py (hunk 24 above)
Comment (ntolley) [reply hnn_core/network.py:571]: Mistake carried over from previous commits that changed the write function. It just looks weird because of the rebase.

**Comment by ntolley** [comment]
That should be everything! Leaving a few conversations unresolved as I think some of the comments will be useful to reference in future PR's.

**Comment by jasmainak** [comment]
It's in, thanks @ntolley ! 🍻 🧠
